### ISSUE TITLE
Rollup + Web Bundles

### ISSUE THREAD

**Issue by hayatoito** [rollup/rollup#4729]
### Feature Use Case

Let me file an issue here to explore an oppornity of Rollup + Web Bundles.

Context: We had on-going discussions based on emails, but it would be better to discuss here to get folks involved from Rollup and related projects there compared to email (and it will serve as a useful discussion to reference in proposals later).

References:
- [Make resource loading better with ESM and Web Bundles](https://docs.google.com/document/d/1ynJTJTdhs1t5awscawO7jomcJwsm_FnSN-qxEvbpqyM/edit?resourcekey=0-i88WuDArpBpCGihCKUuGRw)
- [rollup-plugin-webbundle](https://github.com/GoogleChromeLabs/rollup-plugin-webbundle)
- [Bundle Dependencies](https://github.com/WICG/webpackage/blob/main/extensions/proposals/dependencies-section.md)

### Feature Proposal

- [Make resource loading better with ESM and Web Bundles](https://docs.google.com/document/d/1ynJTJTdhs1t5awscawO7jomcJwsm_FnSN-qxEvbpqyM/edit?resourcekey=0-i88WuDArpBpCGihCKUuGRw) includes "Cache-Aware Bundles" (for dynamic serving use cases in order to reduce the transfer size).

- [Bundle Dependencies](https://github.com/WICG/webpackage/blob/main/extensions/proposals/dependencies-section.md) is a proposal for serving static contents with Web Bundles, including metadata of dependency graphs for better serving.

- Any other proposals, ideas are welcome!

**Comment by hayatoito** [comment]
cc: @patak-dev @lukastaegert just in case.

**Comment by patak-dev** [comment]
Thanks for moving the discussion out of the mail threads @hayatoito 
cc: @danielroe @bluwy

**Comment by hayatoito** [comment]
Let me answer a question (in the email thread) here from @lukastaegert:

> But of course if you think there is enough use case for the static web bundle case, we can talk about everything. One problem Rollup might indeed solve via web bundles is the proliferation of very small chunks you have in large applications. Merging small chunks into few web bundles might solve issues here, as web-bundles are not executed immediately, so we can store arbitrary chunks with side effects together in such a web bundle chunk.

Regarding "as web-bundles are not executed immediately", I've added Q and A (and a related screenshot) to [the doc](https://docs.google.com/document/d/1ynJTJTdhs1t5awscawO7jomcJwsm_FnSN-qxEvbpqyM/edit?resourcekey=0-i88WuDArpBpCGihCKUuGRw).

> Q. Can a browser load subresources in Web Bundles progressively while fetching the bundle?
> A browser can start to load a subresource (and execute it if it is JS) from a webbundle progressively even when the browser doesn't receive the full range of the bundle from the network.

If my understanding of "web-bundles are not executed immediately" is wrong, please correct me. In any case, loading resources progressively from a Web Bundle is one of the important factors, so it would be worth mentioning explicitly here.


> Vite's dev use case on the other hand is very important, as they are currently heavily suffering from request congestion on initial load over http, something which a web-bundle could solve handily.

Probably Vite folks have opinions on this!

**Comment by ArnaudBarre** [comment]
Thanks for investigating this! I just read the first pages, I will read the rest later but I want to point out two things:

For development, the "request congestion" on initial load is not due do the limitation of the browser (when using localhost) but due to the fact that it takes time to _process_ files (transform, parse, resolve and rewrite imports). A way to see this is that once is server is hot, cmd+R is fast despite re-requesting hundreds of modules to the dev-server (because you hit the in-memory cache). Web bundles will probably be an interesting performance bonus, but they will not be a magic solution. 

The P4 and P5 sounds strange:
Even with web bundles, the wide majority of people will still write TS, use syntax lowering, minify their code (for things that gzip can't do like constant inlining), treeshake it (because you can still have dead code inside a given module), ...
Plus once see that esbuild can do all of that and output megabytes "classic" bundles sub second, you know that issue is not the process.

**Comment by hayatoito** [comment]
Thanks for the feedback!

> The P4 and P5 sounds strange: Even with web bundles, the wide majority of people will still write TS, use syntax lowering, minify their code (for things that gzip can't do like constant inlining), treeshake it (because you can still have dead code inside a given module), ... Plus once see that esbuild can do all of that and output megabytes "classic" bundles sub second, you know that issue is not the process.

We might want to involve @Jxck. 

Hi @Jxck, are you comfortable to remove P4 and P5 from there? Probably they can be discussed as another topic. It might be better to focus on P1, P2. and P3 (and any other hidden issues) here.

**Comment by bluwy** [comment]
> For development, the "request congestion" on initial load is not due do the limitation of the browser (when using localhost) but due to the fact that it takes time to _process_ files (transform, parse, resolve and rewrite imports). A way to see this is that once is server is hot, cmd+R is fast despite re-requesting hundreds of modules to the dev-server (because you hit the in-memory cache). Web bundles will probably be an interesting performance bonus, but they will not be a magic solution.

I don't think this is true. There are cases where the Vite dev server is already ready but there are a lot of requests made that it's throttled. My experience so far hasn't been that Vite is taking time to process those files as we made sure there's no parsing involved. Cmd+R is fast because it's already cached by the browser.

---

I think my main concern so far is how HMR will work with Web Bundles. When we invalidate a file or multiple files, we need to re-fetch new ones. If a file is part of a Web Bundle, it might refetch the entire bundle unnecessarily (?). In that case maybe we need to bypass it somehow.

**Comment by ArnaudBarre** [comment]
> we made sure there's no parsing involved

Almost all framework have a toolchain to apply on components files. Plus every file is passed to `es-module-lexer`, which is fast, but not "no parsing".

When doing cmd+r only deps are cached at the browser level, all individuals source files goes through the http dev-server round trip. You can see the waterfall in the browser, but when the dev-servers respond in few millisecond it's not a big issue.

**Comment by bluwy** [comment]
> When doing cmd+r only deps are cached at the browser level, all individuals source files goes through the http dev-server round trip. 

Deps are hard-cached between server starts, but source files are also cached but per server session, so there's still a form of caching from the browser for all files.

**Comment by lemanschik** [comment]
is wip since long time
gets now done but there is something much more big happen that needs to land before be patient. 

if you need a shim able build i just written up something nice https://github.com/lemanschik/modules watch there in rollup rollup.md there are the build instructions you will take one more line for a web build via a codespace from that

**Comment by lemanschik** [comment]
@hayatoito https://docs.google.com/document/d/1ynJTJTdhs1t5awscawO7jomcJwsm_FnSN-qxEvbpqyM/edit?resourcekey=0-i88WuDArpBpCGihCKUuGRw this doc is missing the most importent part the context url based Service Worker which is designed to solve all P* problems programatical

**Comment by hayatoito** [comment]
> @hayatoito https://docs.google.com/document/d/1ynJTJTdhs1t5awscawO7jomcJwsm_FnSN-qxEvbpqyM/edit?resourcekey=0-i88WuDArpBpCGihCKUuGRw this doc is missing the most importent part the context url based Service Worker which is designed to solve all P* problems programatical

Thanks for the feedback.

Could you elaborate how ServiceWorker resolves P* issues? Does the scope include production serving too?
I'm happy to add that as an alternative idea into the doc so that we can compare pros and cons.

**Comment by lemanschik** [comment]
@hayatoito i try but it is hard to do that in short as my conclusions are based on over a decade of ESM raw usage.  And Distributed Engineering in Fact i am working at present on building chromium on the web and also v8 and nodejs distributed via webrtc and not goma!  

that sayed you should maybe look into https://github.com/nodejs/package-maintenance/issues/542

where i have a discussion with one of the most active NPM NodeJS Spec people and he comes out of his view to the conclusion that only CJS makes sense. I understand his point but i take a total diffrent point as i come from the same direction as you.

Ok lets focus on the P* problem and service workers Loading speed is relativ on the internet speed in general is relativ your matrix is not tacking any network distruptions into account and is assuming that everything works optiomal

this is the most rare case as also HTTP3 is not considered which is the new standard. 

ServiceWorkers do offer a URL Context isolated shared Cache so if you cache something in tab 1 on same url you also cache it for tab 2 you can do also cross tab messaging with them.


They solve the problem of relative speed as they can show alternativ data until the real data is there so allow incremental upgrading the app till it is in a usable state. 

You start of with a minimal loader api and then incremental start loading only parts that get next into the viewport. 

So you need to change your overall mind set when packaging and distributing apps take Lighthouse as example and the suggestions it does. Service Workers are able to offer the best User Experience in all cases.

They are a Offline Online Forward Proxy something that we use in companys always for good reasons same reasons do apply here. 

So they solve the problem not out of the box but they give you the backbone to solve it when you design your applications right. 

In general something total new will come up as sayed above in future we will ship more and more cached rollup bundels they are serialized JSON so we can then spit them out directly as CJS ESM iife WORKERS and so on. 

this allows us to share the biggest part of the code and only modifie the export import references as needed.

## Short for your document
Speed is relative the Matrix can be thrown away in real world Service worker do solve that via offering the Application author to controls what loads when from where and how it gets cached. 

example for relativ UX speed even on Local Host:

in VSCODE when you switch the tabs between 2 large documents the tab slider at the top switches directly but not the window main content this lets the user think that it is snappy while it is not. 

i guess thats a good example for relativ speed feeling.

**Comment by lukastaegert** [comment]
@hayatoito What I mean with "as web-bundles are not executed immediately" is the following, and please correct me if I am wrong, following one of your documents:

- The browser parses a web bundle script tag
  ```html
  <script type="webbundle">
  {
    source: "a.wbn",
    resources: ["a1.js", "a2.js"]
  }
  </script>
  ```
- at this point, it will start loading `a.wbn`, but it will not execute any of its resources, right? So we can put in anything we want, even stuff that the page does not need right away or at all.
- However it creates like an internal "import map" that ensures all subsequent imports of `a1.js` and `a2.js` will go against the web bundle and not trigger additional downloads.
- Only when there is another script tag after the web bundle
  ```html
  <script type="module" src="a1.js"></script>
  ```
  anything will actually be executed.

So I keep thinking about how this fits with Rollup. My first impulse is that web bundles are mainly useful for higher level tooling, and I think its true power is with just-in-time cache-aware web bundle creation where
- the browser tells the server which resources are already cached (including e.g. etags to identify resource versions)
- the server assembles all the missing resources into a single bundle

This would in theory work quite well for initial load requests.

Now my understanding gets sketchy when it comes to handling dynamic imports. At this point, we want to declare "lazy" dependencies in the web bundle, which apparently we can, so that when a dynamic import occurs, we immediately request a bundle for the import target. If again the browsers sends along everything they already have as cached resources, this bundle can be tailored to only contain the resources they need.

If your strip away the dynamic server from this equation and just build static bundles that contain all the files, then this is also something that Rollup could do.

But without the dynamic server, you either get the "too many small chunks" problem, that Rollup already has when there are many dynamic imports, or massive over-fetching when you try to solve this problem by putting the same resources into different chunks to avoid the waterfall (but I guess this is not even possible, i.e. what would the browser do if I say the dynamic resource `d.js` is in bundle `a.wbn`, but at a later time I load bundle `b.wbn` which also contains `d.js`? Throw or ignore or something else?). Cache-aware JIT rendered bundles could solve this issue by omitting `d.js` from the second bundle.

There are many more thoughts in my head, but this is a start.

**Comment by lemanschik** [comment]
you will soon see something in action on a bigger project i guess i try to port over some knowleg to the engineer that does the ESM Migration of vscode: https://github.com/microsoft/vscode/issues/167505

**Comment by lemanschik** [comment]
the nice about the above referenced issue is it will lead to in memory code deduplication by x * x as also will open incremental adoption of JSON serialized cached Unified Import information. to create ESM CJS out of In memory cache on the fly.

**Comment by lemanschik** [comment]
here i will show the concept via the node cli i call it node intermmediate result https://github.com/lemanschik/nodeir

**Comment by lemanschik** [comment]
Oh ya and in fact all above sayed takes production directly into account i used the following pattern since first inital definition 2014! 

- transpil all dependencies up ! 
- create shareable rollup dev bundle
- compose rollup dev bundels
- create new virtual entrys to export only whats needed
- share that in memory cross context.

**Comment by hayatoito** [comment]
> @hayatoito What I mean with "as web-bundles are not executed immediately" is the following, and please correct me if I am wrong, following one of your documents:
>   ... (omit)
>     
>   anything will actually be executed.

That's exactly correct. Thanks for clarifying it.

> 
> So I keep thinking about how this fits with Rollup. My first impulse is that web bundles are mainly useful for higher level tooling, and I think its true power is with just-in-time cache-aware web bundle creation where
> 
> * the browser tells the server which resources are already cached (including e.g. etags to identify resource versions)
> * the server assembles all the missing resources into a single bundle
> 
> This would in theory work quite well for initial load requests.

Yes, that matches what we call "Cache-Aware Bundles" (there are small differences, but that doesn't matter for now)

> 
> Now my understanding gets sketchy when it comes to handling dynamic imports. At this point, we want to declare "lazy" dependencies in the web bundle, which apparently we can, so that when a dynamic import occurs, we immediately request a bundle for the import target. If again the browsers sends along everything they already have as cached resources, this bundle can be tailored to only contain the resources they need.

How "Cache-Aware Bundles" can handle dynamic imports would be an interesting problem. I'd like to hear opinions from @dvoytenko.
At first thought, inserting `<script type=bundlepreload>` element dynamically into the DOM tree should work, though the script element should be surely inserted just before actual dynamic import execution happens in JS).

> 
> If your strip away the dynamic server from this equation and just build static bundles that contain all the files, then this is also something that Rollup could do.
> 
> But without the dynamic server, you either get the "too many small chunks" problem, that Rollup already has when there are many dynamic imports, or massive over-fetching when you try to solve this problem by putting the same resources into different chunks to avoid the waterfall (but I guess this is not even possible, i.e. what would the browser do if I say the dynamic resource `d.js` is in bundle `a.wbn`, but at a later time I load bundle `b.wbn` which also contains `d.js`? Throw or ignore or something else?). 

In the current implementation, we have a simple rule: the firstly parsed script element wins. `d.js` in `b.wbn` is ignored.
I have to confess I have to check Chromium's source code because I don't remember this rule.

For example (this is not dynamic loading case, but it is the same):

```
<script type="webbundle">
{
  source: "a.wbn",
  resources: ["d.js"]
}
</script>
<script type="webbundle">
{
  source: "b.wbn",
  resources: ["d.js"]
}
</script>
<script type="module" src="d.js"></script>
```

Note you can remove the script element of `a.wbn` from the DOM tree to remove the `import-map` of the element, After that, `d.js` should be loaded from `b.wbn`.

> There are many more thoughts in my head, but this is a start.

Thanks! Regarding static bundles and dynamic imports, I thought that a little about one year ago and wrote this: [Bundle Dependencies](https://github.com/WICG/webpackage/blob/main/extensions/proposals/dependencies-section.md).

This is just a rough idea, but what I want to emphasis here is "anything is possible" (at least in this brainstorming stage). I'd like to hear any ideas, proposals, any feature requests, or new API. Thanks for sharing your thoughts quickly!

**Comment by lemanschik** [comment]
@hayatoito would like to work with you together more directly you can write it up technical and i can deliver you the needed information in fact i am working on that since 2014 and did a lot of research experiments as also can supply greenfield patterns. 

but some one needs to bring all that into shape it is a lot because everything depends on everything this everything goes as deep that we even experiment with building v8 inside v8

**Comment by lemanschik** [comment]
I even work on running v8 as kernel ring 0 level

**Comment by lemanschik** [comment]
## example greenfield imports 

```js
import('vscode-resolve?id=vs/editor').then( ({ resolvedId }) => import(resolvedId) );
```

**Comment by lemanschik** [comment]
at present the final goal would be a webrtc shared global goma like build grit in memory with a couchdb like sync 
and in fact this is not only a idea this is total do able..... we could not share a package registry we could share directly like a blockchain incremental builds and reference them

**Comment by lemanschik** [comment]
in fact package managers today are switching to content hashed caches anyway using sha 512 so we can build up on that

**Comment by lemanschik** [comment]
only for your interrest as your googler using puppeteer to spin up a page using this tech using CDP the --remote-debbuging-pipe which allows stdio pipe to devtools allows us to link up nativ build components in a module like fashion like a userland implementation of nativMessaging api

i use that to run the monaco editor on demand inside devtools using overrides and workspaces as also directly the chromium devtools layout abilitys to create dev environments.

**Comment by lemanschik** [comment]
i mean to make it short lets directly build a distributed p2p dag graph content addressed network of the main repos fuchsia chromium android and call it a day?

**Comment by patak-dev** [comment]
@lemanschik these comments are completely off-topic for this discussion, please create a discussion in your own repo and avoid further messages here. You are creating tons of notifications for the maintainers involved in this issue. Thanks!

**Comment by lemanschik** [comment]
i am in fact rumping that up including a web installer at https://github.lemanschik.com/src_chromium/

i use ghpages with a nojekyll to not crash github with html creation and share the chromium src import able as ESModules

**Comment by lemanschik** [comment]
@patak-dev it makes the complet web bundel stuff obsolet how can that be unrelated?

**Comment by lemanschik** [comment]
Who wants to bundle when he can directly run his code? thats the whole core concept of the v8 vm to JIT Code!

**Comment by lemanschik** [comment]
The Fundamentals are not code bundle deploy blub bla bla bla add instrumentation blub bla bla add observability blub bla bla monitor blub bla bla 

we want to run code! 

do never forget that we code => run 

code is just a language for instructions with the lead to run not to get bundled.

**Comment by lemanschik** [comment]
Lets never forget the bundle deploy stuff is a sympthom of wrong old patterns which we can avoid today. We can bring JS back to what it was a nice expressiv language to define Streams so Connections between Components in a dynamic way today even with lazy static typing ability

**Comment by lemanschik** [comment]
then we can connect all kind of stuff in memory wasm, nativ components, ESModules that contain connections and logic for handling events.

**Comment by lemanschik** [comment]
look at gulp its a task runner it evolved into what? a streams lib look what we got in the web Readable writeable streeams again right we can reduce that its simple math we can drop over 80% and maybe even more of over 50 mio Codebases in open source space

**Comment by lemanschik** [comment]
it should never be the goal to build the next framework ECMAScript got created with a total other goal it should connect components and allow you to do some logic like LUA we got a API state we agreed on and that is more then enough to drop bundling while resolve stays usefull.  The ECMAScript host is in Fact a JIT Compiler

**Comment by lemanschik** [comment]
if you build v8 for a platform arch you did build a whole Operating system that is script able.

**Comment by lemanschik** [comment]
Android does nothing other it takes the JVM but this can get replaced by V8 v8 is GraalVM Equal!!!!

https://www.graalvm.org/

**Comment by lemanschik** [comment]
even docker is not needed anymore whole lib container you can throw that away.

**Comment by lemanschik** [comment]
you can throw kubernets away

**Comment by lukastaegert** [comment]
I am sorry Frank but this issue is solely about the usage of web bundles together with Rollup. If you have plans to fix the Web that is fine, but please talk about them elsewhere. If you keep abusing this channel, I will block you.

**Comment by dvoytenko** [comment]
@hayatoito the first thing I'd like to confirm: do you intend to use "Resource Subsetting" option of Web Bundles here? If yes - that'd be a version of dynamic bundling that's mentioned in other comments.

I also don't see how P4/P5 are helped by Web Bundles. Roughly, one plausible way I see how Rollup and Web Bundles could work is this:

Let's take a moderate application with 10,000 source modules. These modules import each other and the dependency graph has arbitrary depth. Rollup should still be used as today: rebundle the source modules, inline functions, perform DCE, etc - all around a set of entry points. The entry points can be established around routes, major components, and dynamic imports. Let's say the result is a set of 300 output modules and the dependency graph is of limited depth (maybe 3-4. E.g. a route entry point imports a feature that imports a shared component, etc).

Rollup already does all of these tasks very well, with an option of creating the output modules as ESM, or CommonJS, or UMD, and more. However, serving these 300 modules could still be expensive for two reasons:
1. Even with limited depth, imports cause waterfall dependency loading.
2. Downloading N dependent modules via N requests is more expensive than a single request: (a) compression is less effective, and (b) a lot of per request/response overhead.

A server could be configured to send `modulepreload` for all dependencies in depth when an EMS module is requested. This could reduce the problem (1), but wouldn't help (2).

Thus Rollup does a great job of rebundling modules. Web Bundles (with resource subsetting) could help serving them. For instance, if a browser is aware that modules can be served via Web Bundles and has some knowledge of the dependency graph, it could bunch up several requests/responses together and thus address both (1) and (2) above in a cache-aware way. But that implies that the server need to be able to serve dynamic subsets of modules, not a single static Web Bundle.

**Comment by hayatoito** [comment]
> @hayatoito the first thing I'd like to confirm: do you intend to use "Resource Subsetting" option of Web Bundles here? If yes - that'd be a version of dynamic bundling that's mentioned in other comments.

Yes, roughly we have two use cases, dynamic bundling (Resoruce Subsetting, aka Cache-Aware Web Bundles) and static bundling (which doesn't require smart server). Both are important for me. Eventually, we might want to split this thread into each one.

> I also don't see how P4/P5 are helped by Web Bundles.

Thanks! I've removed P4 and P5 to avoid the confusion (with a permission of @Jxck).

> Roughly, one plausible way I see how Rollup and Web Bundles could work is this:

Thanks for the explanation. I totally agree. "pre-bundle" task is required for several important reasons.
(e.g. 3000 modules -> 300 modules for very good reasons). I don't have any intention to remove this "pre-bundle" stage.
My primary goal is to resolve "Production Serving" problem, which Web Bundles could help.

**Comment by lukastaegert** [comment]
As my understanding grows, this is how I think how web bundles could play out with Rollup/Vite:
- Even if you intend to use static web bundles, you would still want Rollup create all regular JS assets and also deploy them with your site as a fallback for browsers not supporting web bundles.
- That means web bundle creation would in any case be an extra step on top. It could be done by a Rollup plugin, but it could also be done by a Vite plugin.
- Using a Vite plugin has the advantage that the plugin could also take care of injecting the web bundle definition into the HTML.
- As an initial improvement and since we have the regular files as fallback, one could start with a mixed approach that would also avoid over-fetching: Each HTML-page only gets a single web bundle which contains exactly the files (including transitive dependencies and other resources) needed for the initial load.
- After that, dynamic imports will just load fallback files.

This would already be a huge win for SEO as the initial load is the most critical, and could be an interesting MVP to build upon.

**Comment by dvoytenko** [comment]
@lukastaegert I think this approach makes sense. The one thing that it wouldn't quite cover though (or not very well) are dynamic imports. They'd either need to be overpacked into static bundles or be served less efficiently via ESM waterfall.

**Comment by hayatoito** [comment]
Thanks. I am always a fan of an incremental approach and want to have a MVP.
I'd never thought this issue from perspective of SEO. That's really helpful!

It's still a bit unclear how MVP can be, but I hope we can have it once we can have more concrete ideas. Thanks for insightful feedback!

**Comment by lukastaegert** [comment]
> or be served less efficiently via ESM waterfall.

Exactly. At least I do not see how we can improve very much here with static web bundles considering we do not know the order in which dynamic imports are triggered. But of course that is why I propose this as an MVP to get out something usable.

The far goal would be a solution with dynamic bundles, but a solution with static bundles would also help to gain some learnings. One small improvement could be to put resources like CSS fragments into web bundles with the corresponding JavaScript chunks. But again that would be something Vite would know more about considering they are doing the CSS handling.

> It's still a bit unclear how MVP can be

Considering there is already https://github.com/GoogleChromeLabs/rollup-plugin-webbundle to take inspiration from, this could be extended into a `vite-plugin-webbundle` that does the following:

- for each entry chunk that corresponds to an HTML page, identify all static dependencies and put them together into a web bundle placed next to the entry chunk.
- use `transformIndexHTML` to inject the web bundle script tag before the entry script tag

And I guess that should already be it for having something useable and already useful. In subsequent steps, one could look into integrating this with CSS handling and possibly other resources.

Once we have that, one could start thinking about something like `vite-plugin-webbundle-server` to create an actual (Node?) server together with the output to generate dynamic bundles (a question here would also be if this should be combined with existing SSR solutions, or again if the first version would be without SSR integration to have a vanilla version and later see how this can be combined with the different SSR approaches per framework).

**Comment by lukastaegert** [comment]
In parallel, one could look in how far web bundles really can help Vite startup performance. One advantage I see is since web bundles can be streamed, the dev server can already start sending parts of the web bundle even though the server is not yet done transforming all files/loading them from some cache, and the browser can already start parsing. If the files in the web bundle are carefully ordered, I would expect a huge performance boost from this.

There was one question

> I think my main concern so far is how HMR will work with Web Bundles. When we invalidate a file or multiple files, we need to re-fetch new ones. If a file is part of a Web Bundle, it might refetch the entire bundle unnecessarily

I do not think this is how HMR works (but I am not the expert myself, admittedly). You cannot just invalidate and replace some ES module in a browser as the module graph can only be extended at runtime, at least I am not aware of any API that does that. Instead, HMR requires your modules to be "instrumented" in some way so that their content can be replaced at runtime with the content of another file. The new file is then loaded using a custom request parameter, I think for Vite it is something like `/fileName.js?t=<timestamp>`, which does not really care if the original file was part of a web bundle. So having the first request served from a web bundle should not have any impact on replacing individual files via HMR later.

**Comment by lemanschik** [comment]
@lukastaegert the solution for that thing is throwing away the whole document context the module graph is assigned to

in general the most best is to create a streaming document then load the modules inside that with its references 
then throw away the doc if you want to re generate the module graph 

https://developer.mozilla.org/en-US/docs/Web/API/DOMImplementation/createHTMLDocument

its a none widly known api most time only used by implementers this is also the solution to streaming rendering while it is bad to block the rendering th this does not do so so you can stream documents as it remembers the parsers open tag.

it is a indipendent html parser including dom including module graph

```js
 let doc = document.implementation.createHTMLDocument("my-web-bundle-hash");
```

**Comment by lemanschik** [comment]
> References:
> 
> * [Make resource loading better with ESM and Web Bundles](https://docs.google.com/document/d/1ynJTJTdhs1t5awscawO7jomcJwsm_FnSN-qxEvbpqyM/edit?resourcekey=0-i88WuDArpBpCGihCKUuGRw)
> * [rollup-plugin-webbundle](https://github.com/GoogleChromeLabs/rollup-plugin-webbundle)
> * [Bundle Dependencies](https://github.com/WICG/webpackage/blob/main/extensions/proposals/dependencies-section.md)

**Comment by bluwy** [comment]
Sorry things have been catching up with me 😅 

> I do not think this is how HMR works (but I am not the expert myself, admittedly). You cannot just invalidate and replace some ES module in a browser as the module graph can only be extended at runtime, at least I am not aware of any API that does that. Instead, HMR requires your modules to be "instrumented" in some way so that their content can be replaced at runtime with the content of another file.

Yup Vite is currently doing so with the [`import.meta.hot` API](https://vitejs.dev/guide/api-hmr.html) for framework authors to declare how to invalidate the old module and how to get replaced by the new incoming module.

> The new file is then loaded using a custom request parameter, I think for Vite it is something like `/fileName.js?t=<timestamp>`, which does not really care if the original file was part of a web bundle. So having the first request served from a web bundle should not have any impact on replacing individual files via HMR later.

Yeah Vite does [attach a timestamp](https://github.com/vitejs/vite/blob/a08ca078515b1f8b33eeb193d0c18eca05811271/packages/vite/src/client/client.ts#L426-L433) when re-importing it. I didn't know that a timestamp query would bypass the web bundle so that's sounds like a non-problem for Vite now! In hindsight it probably make sense but I didn't connect before.

**Comment by hayatoito** [comment]
FWIW, Google will step away from doing active exploration of new features of Web Bundles, as mentioned in [Make resource loading better with ESM and Web Bundles](https://docs.google.com/document/d/1ynJTJTdhs1t5awscawO7jomcJwsm_FnSN-qxEvbpqyM/edit?resourcekey=0-i88WuDArpBpCGihCKUuGRw), this year.

So I'm afraid that I couldn't contribute this thread any more until the situation has changed.

Thanks for giving us many advises and insights in this issue. I really appreciate them. I hope we can work together someday.
